src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -799,6 +801,23 @@ private void GetReadResult(out ReadResult result)
             }
         }
 
+        private void SetMemory()
+        {
+            // REVIEW: Optimize the single block case?
+            BufferSegment node = _readHead;
+
+            while (true)
+            {
+                node.UpdateMemory();
+
+                if (node == null || node == _readTail)
```

Comment:
Is checking `node == null` after `node.UpdateMemory();` neccessary?

Reply:
Nope, that could be moved to the loop condition. The tail check has to be done after but the null check could be moved.

Reply:
If this version of SetMemory() isn't broken, the null check isn't necessary at all. If it was, this would null ref. Seems to me that `do { node.UpdateMemory(); node = node.NextSegment; } while (node != _readTail);` is functionally equivalent to this.

Reply:
My above do/while loop is wrong since you need to test `node != _readTail` before advancing to the next segment, but I still don't think you don't need the null check.